### PR TITLE
Jfm merge across channels

### PR DESCRIPTION
**Pull request by magland**
Implemented mscmd_merge_across_channels

### CODE REVIEW

File: mountainsort/src/isosplit/isosplit2.cpp [hunk 1]
```diff
@@ -63,28 +63,28 @@ void geometric_median(int M,int N,double *ret,double *X) {
...
-QList<double> compute_center(Mda &X,const QList<int> &inds) {
+QList<double> compute_center(Mda &X,const QList<long> &inds) {
     int M=X.N1();
     int NN=inds.count();
     if (NN==0) {
         QList<double> ret; for (int i=0; i<M; i++) ret << 0;
         return ret;
     }
-	double *XX=(double *)malloc(sizeof(double)*M*NN);
+    double *XX=(double *)malloc(sizeof(double)*M*NN);
```
Comment (wysota) [inline mountainsort/src/isosplit/isosplit2.cpp]: Avoid using malloc in C++ code.
```
double *XX = new double[M*NN];
```

File: mountainsort/src/isosplit/isosplit2.cpp [hunk 2]
```diff
@@ -199,146 +199,146 @@ void find_next_comparison(int M,int K,int &k1,int &k2,bool *active_labels,double
...
-	}
-	return ret;
+        double val=0;
+        for (int m=0; m<M; m++) val+=X2.value(m,i)*V[m];
+        XX << val;
+    }
+
+    QList<double> XXs=XX;
+    qSort(XXs);
+    double *XXX=(double *)malloc(sizeof(double)*(N1+N2));
```
Comment (wysota) [inline mountainsort/src/isosplit/isosplit2.cpp]: Avoid malloc. Use std::vector or any of the other structures mentioned earlier.

File: mountainsort/src/isosplit/isosplit2.cpp [hunk 3]
```diff
@@ -199,146 +199,146 @@ void find_next_comparison(int M,int K,int &k1,int &k2,bool *active_labels,double
...
-		}
-	}
-	return ret;
+        double val=0;
+        for (int m=0; m<M; m++) val+=X2.value(m,i)*V[m];
+        XX << val;
+    }
+
+    QList<double> XXs=XX;
+    qSort(XXs);
```
Comment (wysota) [inline mountainsort/src/isosplit/isosplit2.cpp]: Is "XXs" used anywhere?

**Comment by wysota** [comment]
In many places you are using `QList`. While with some of the them it doesn't matter, with others it would be more efficient* to use `QVector`. Since you are only appending to the structure, there is completely no benefit from using `QList` over using `QVector`.

*) efficiency rarely comes either from using a different data structure directly; most often you can simply use faster algorithms on a vector compared to a list (which doesn't have some of the guarantees which are true for a vector). I would strongly advise to review the code (and API) and use vectors where applicable.

An example benefit is that in some situations you are initializing a list with 0 values by appending 0 multiple times. With a vector you could do that in one go more efficiently. You could also use functios such as memcpy() (or std::copy) to quickly copy content between instances. Finally you could use some SIMD instructions which could make many of the operations that you do in the processor significantly faster. If you are interested in the latter, drop me an email and I will point you to a presentation on it.

**Comment by magland** [comment]
Thank you. I would like to use the appropriate container throughout the entire project.

The thing is, I don't like to mix QVector and QList because then I have to convert between them. This is why I settled on QList for everything. But maybe I should be using QVector for everything. Or.... maybe I should use QVector for arrays of numbers and QList for everything else (like strings or structs).

I need a clear rule on when to use one or the other, so that we can modify all the files in the project in one go.

**Pull request merged**